### DISCUSSION TITLE
default export not working on pure-ESM typescript (node)

### DISCUSSION THREAD

**Discussion by Polve** [node-fetch/node-fetch#1642] [Q&A]
I'm developing a new project so I can start from scratch using latest standards, so I'm using latest 16.x node, latest typescript (4.8.2), and trying to develop an ESM library.

The problem I have is that I can't import the default export, while everything works fine for the named imports.

So for example this works fine:
```js
import {  Blob,  File,  Headers,  Request,  Response } from 'node-fetch'
```
while this code:
```js
import fetch from 'node-fetch'
```
gives this error:
```
/home/user/.nvm/versions/node/v16.13.0/lib/node_modules/ts-node/dist/index.js:851
            return old(m, filename);
                   ^
Error [ERR_REQUIRE_ESM]: require() of ES Module /home/user/x/node_modules/node-fetch/src/index.js from /user/x/foo.ts not supported.
Instead change the require of index.js in /home/user/x/foo.ts to a dynamic import() which is available in all CommonJS modules.
    at Object.require.extensions.<computed> [as .js] (/home/x/.nvm/versions/node/v16.13.0/lib/node_modules/ts-node/dist/index.js:851:20)
```

Is it possible for you do avoid exporting the default and simply do a named export for "fetch"?
Or maybe both, to have better compatibility with cjs?

I'm no export on ESM modules but I think that a trivial fix would be to avoid exporting the definition of the function this way:
```js
export default async function fetch(url, options_) {
```
and define the function and export it in two different places like this:
```js
async function fetch(url, options_) {...}
export default fetch;
```

**Comment by jimmywarting** [comment]
> `Error [ERR_REQUIRE_ESM]: require() of ES Module /home/user/x/node_modules/node-fetch/src/index.js`

implies to me that you compile your typescript to commonjs?

would it be possible for you to compile to ESM instead?

**Comment by Polve** [comment]
Compile to ESM is my goal, but I'm not sure I'm doing it right.

in my `tsconfig.json` file I have:
```json
"module": "NodeNext"
"moduleResolution": "node"
"esModuleInterop": true
```
while in my `package.json` I've *no* special option set (so I'm *not* using  `"type": "module"`)

**Reply by jimmywarting** [reply]
could you create a minimal reproducable example over at https://stackblitz.com/ ?

**Reply by LinusU** [reply]
> while in my package.json I've no special option set (so I'm not using "type": "module")

I think this is why it isn't working then. As I remember, if you set `"module": "NodeNext"` but not `"type": "module"` in package.json, TypeScript will output CommonJS code.

Set `"type": "module"` in your package.json and you should be good to go...

**Comment by Polve** [comment]
Actually after tinkering a bit I discovered I was wrong: there is the same problem using the named imports.
The problem arises only if I *use* the imported symbols, otherwise everything *seems* to work fine.

So now I'm stuck again, and I can't understand how to develop a pure ESM module in typescript :-|

**Reply by GabenGar** [reply]
There is no such thing as "pure ESM module in typescript" as of now. A lot of types pre-NodeESM were written for a typescript implementation of ESM, which was pretty liberal at interpreting them.

**Comment by mannymu** [comment]
I have the same problem, has anyone solved it?

**Comment by Polve** [comment]
I solved it by using node 18+ that has integrated fetch :-)
